### PR TITLE
tests.cuda.cuda-{,library-}sample_*: Init at supported CUDA toolkit versions

### PR DESCRIPTION
**Pull request by Ericson2314**
###### Motivation for this change

Since CUDA is unfree, we won't actually use this when testing Nixpkgs
officially. But I want to include this as they are useful for users of
Nixpkgs trying to set up / debug a CUDA environment.

###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [ ] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/test/default.nix [hunk 1]
```diff
@@ -47,5 +47,7 @@ with pkgs;
 
   texlive = callPackage ./texlive {};
 
+  cuda = callPackage ./cuda {};
```
Comment (SuperSandro2000) [inline pkgs/test/default.nix:50]:
```suggestion
  cuda = callPackage ./cuda { };
```

File: pkgs/test/cuda/cuda-samples/generic.nix [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+
+    install -Dm755 -t $out/bin bin/${stdenv.hostPlatform.parsed.cpu.name}/${stdenv.hostPlatform.parsed.kernel.name}/release/*
+
+    runHook postInstall
+  '';
+
+  meta = {
+    description = "Samples for CUDA Developers which demonstrates features in CUDA Toolkit.";
+    # CUDA itself is proprietary, but these sample apps are not.
+    license = lib.licenses.bsd3;
```
Comment (SuperSandro2000) [inline pkgs/test/cuda/cuda-samples/generic.nix:42]: I think it wouldn't hurt if you added yourself as a maintainer.

File: pkgs/test/cuda/cuda-samples/generic.nix [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+  installPhase = ''
+    runHook preInstall
+
+    install -Dm755 -t $out/bin bin/${stdenv.hostPlatform.parsed.cpu.name}/${stdenv.hostPlatform.parsed.kernel.name}/release/*
+
+    runHook postInstall
+  '';
+
+  meta = {
+    description = "Samples for CUDA Developers which demonstrates features in CUDA Toolkit.";
```
Comment (SuperSandro2000) [inline pkgs/test/cuda/cuda-samples/generic.nix:40]:
```suggestion
    description = "Samples for CUDA Developers which demonstrates features in CUDA Toolkit";
```

File: pkgs/test/cuda/cuda-samples/generic.nix (hunk 2 above)
Comment (Ericson2314) [reply pkgs/test/cuda/cuda-samples/generic.nix:42]: I am not sure who is going to end up maintaining this yet, so I'm going to mark it as WIP.

**Comment by Ericson2314** [comment]
WIP pending figuring out maintainer for this.

**Comment by Ericson2314** [comment]
I added some more tests in the meantime

**Comment by knedlsepp** [comment]
All of these need addOpenGLRunpath, right?
(Currently on my phone, so cannot check that but I do remember that also some samples are shipped with the cudatoolkit installation itself)

File: pkgs/test/cuda/cuda-samples/generic.nix [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+  src = fetchFromGitHub {
+    owner = "NVIDIA";
+    repo = pname;
+    rev = "v${version}";
+    inherit sha256;
+  };
+
+  nativeBuildInputs = [ pkg-config ];
+
+  propagatedBuildInputs = [ cudatoolkit ];
```
Comment (knedlsepp) [inline pkgs/test/cuda/cuda-samples/generic.nix:23]: Why propagated? Shouldn't that be a regular nativeBuildInput?

Comment (Ericson2314) [reply pkgs/test/cuda/cuda-samples/generic.nix:23]: Yes I suppose it is just exes.

**Comment by Ericson2314** [comment]
@knedlsepp I don't think so? I've had some issues with the cuda-library-samples ones, but on both RHEL and nixos they ran with an LD_LIBRARY_PATH for NVidia proprietary things, not OpenGL.

**Comment by knedlsepp** [comment]
> @knedlsepp I don't think so? I've had some issues with the cuda-library-samples ones, but on both RHEL and nixos they ran with an LD_LIBRARY_PATH for NVidia proprietary things, not OpenGL.

The cuda libraries are put into `/run/opengl-driver/lib` on NixOS, which makes it possible to run software relying on CUDA without touching the LD_LIBRRARY_PATH. As for most things nixpkgs there's no right answer, but you can find that pattern of using the addOpenGLRunpath for most of th packages. (Some incorrectly link against a hard-coded linuxPackages.nvidia_x11, which doesn't guarantee it matches the version installed globally on your system and then break. I've been trying to get a PR ready to get rid of that. )
As for running on non-NixOS I've had good success with replicating the NixOS way und Putting the libcuda into the same directory.
(I want nix binaries to "just work" on my HPC system without users having to mess with LD_LIBRARY_PATH, which has caused me pain more often than not)

All of this works so well that I'd even be in favour of adding a hook to nixpkgs so that all programs that have cudatoolkit as a nativeBuildInput should receive the /run/opengl-driver/lib rpath similar to autopatchelfhook as soon as they have libcuda.so.1 somewhere in the binary (to catch all cases of dlopen)

**Comment by Ericson2314** [comment]
OK thanks for the info, I'll amend.

> All of this works so well that I'd even be in favour of adding a hook to nixpkgs so that all programs that have cudatoolkit as a nativeBuildInput should receive the /run/opengl-driver/lib rpath similar to autopatchelfhook as soon as they have libcuda.so.1 somewhere in the binary (to catch all cases of dlopen)

you could make the setup hooks `propagatedNativeBuildInputs`; that is for cudatoolkit being a build input which it usually is? (It would also be nice to use multiple outputs to separate tools and libraries).

**Comment by Ericson2314** [comment]
FYI have more cuda stuff in https://github.com/NixOS/nixpkgs/pull/116666

**Pull request merged**